### PR TITLE
feature(gry): add tutorial for RM

### PR DESCRIPTION
**Pull request by ruoyuGao**

### CODE REVIEW

File: source/04_best_practice/reward_model_zh.rst [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+在 DI-engine 中，Reward Model是我们提供的一个组件。所有的Reward Model类都继承自名为\ **BaseRewardModel**\的一个抽象基类。在这个类中，我们定义了最基本的Reward Model功能如下。
+
+.. code-block:: python
+    
+   class BaseRewardModel(ABC):
+
+        @abstractmethod
+        def estimate(self, data: list) -> Any:
+            """
+            给出估计的奖励值
```
Comment (kxzxvbk) [inline source/04_best_practice/reward_model_zh.rst:20]: 这里的注释我感觉可以写的详细些，比如这个函数会修改data中的reward值为reward model给出的值，要注意此方法会导致原始data的reward值无法回复，需要在使用时格外注意

File: source/04_best_practice/reward_model_zh.rst [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    # 你所要训练的main config， create config
+    # cooptrain_reward = True 表示在训练policy的时候，同时训练reward model
+    # pretrain_reward = False 表示在开始训练前，不预训练reward model
+    # 需要注意的是，如果想要使用pretrain_reward，需要在init reward model的时候准备好train data
+    # 目前只有trex支持pretrain_reward
+    serial_pipeline_reward_model_offpolicy((main_config, create_config), cooptrain_reward = True, pretrain_reward = False)
+
+
+
+**在强化学习训练中添加Reward Model**
```
Comment (kxzxvbk) [inline source/04_best_practice/reward_model_zh.rst:64]: 我认为接下来的几段缺了一个总括性的概述，比如我认为应该加一个：如何在强化学习中使用Reward model（类似这样的大标题）。这一标题下分出几点，分别是：如何定义和添加reward model, 如何训练reward model，如何使用reward model预测reward

File: source/04_best_practice/reward_model_zh.rst [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+===============================
+
+Reward Model 入门
+-------------------------------
+
+**Reward Model 的基本概念**
+
+在强化学习中，Reward Model 是对智能体的行为进行评价的模型，它的输入是智能体的观测和动作，输出是一个标量的奖励值。
+
+在 DI-engine 中，Reward Model是我们提供的一个组件。所有的Reward Model类都继承自名为\ **BaseRewardModel**\的一个抽象基类。在这个类中，我们定义了最基本的Reward Model功能如下。
```
Comment (kxzxvbk) [inline source/04_best_practice/reward_model_zh.rst:11]: 英文和中文之间应当存在空格，这个涉及许多地方，都需要修改下

File: source/04_best_practice/reward_model_zh.rst (hunk 2 above)
Comment (kxzxvbk) [inline source/04_best_practice/reward_model_zh.rst:64]: 我认为接下来的几段缺了一个总括性的概述，比如我认为应该加一个：如何在强化学习中使用Reward model（类似这样的大标题）。这一标题下分出几点，分别是：如何定义和添加reward model, 如何训练reward model，如何使用reward model预测reward

File: source/04_best_practice/reward_model_zh.rst [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    # estimate函数会返回一个 list[Dict],其中每个Dict都至少包含obs, action, reward, next_obs, done
+    # estimate会修改传入的train_data的reward为Reward Model预测的reward
+    train_data_augmented = reward_model.estimate(train_data)
+    # 将预测好的reward加入到我们RL算法的训练过程当中
+    learner.train(train_data_augmented, collector.envstep)
+
+如何添加新的Reward Model
+-------------------------------
+
+在上一节中，我们介绍了如何使用现有的reward model。接下来，我们将展示如何添加新的reward model，并且需要遵循哪些规范。
```
Comment (kxzxvbk) [inline source/04_best_practice/reward_model_zh.rst:112]: reward model是否要全文保持首字母大写，需要统一一下

File: source/04_best_practice/reward_model_zh.rst [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+**RewardModel方法介绍**
+
+首先你的reward model需要包含以下方法，具体方法的介绍如下。
+
+.. code-block:: python
+    
+    def __init__(self, config: EasyDict, device: str, tb_logger: 'SummaryWriter') -> None: 
+    """
+    初始化RM，会在create RM的时候调用，需要注意的是如果要导入expert data（可以写一个self.load_expert_data()的方法）
```
Comment (kxzxvbk) [inline source/04_best_practice/reward_model_zh.rst:128]: 这个地方我认为就讲必须初始化的属性吧，如cfg, reward_model, tb_logger。像计数器之类的，由于每个算法不同，也不一定存在，让用户自行定义即可

File: source/04_best_practice/reward_model_zh.rst [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    """
+    pass
+
+    def train(self) -> None:
+    """
+    这个方法的在entry的inner loop中调用，
+    功能是train整个RM，并向logger添加内容，形式应该如下
+    1. 由内部方法_train()进行具体训练，接受需要加入logger的返回值
+    2. 将对应内容添加到logger
+    for _ in range(self.cfg.update_per_collect):
```
Comment (kxzxvbk) [inline source/04_best_practice/reward_model_zh.rst:168]: example:

File: source/04_best_practice/reward_model_zh.rst [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        @abstractmethod
+        def train(self, data) -> None:
+            """
+            训练Reward Model
+            """
+
+        @abstractmethod
+        def collect_data(self, data) -> None:
+            """
+            收集RM所需要的训练数据
```
Comment (PaParaZz1) [inline source/04_best_practice/reward_model_zh.rst:32]: 收集数据好像有点歧义，不如说将外部传入的训练数据存储在 RM 内部用于训练

File: source/04_best_practice/reward_model_zh.rst [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+
+        @abstractmethod
+        def clear_data(self, iter: int) -> None:
+            """
+            清理Reward Model类buffer里的数(即collect_data收集的数据)
+            """
+
+        def load_expert_data(self, data) -> None:
+            """
+            加载专家数据，只有在使用专家数据训练Reward Model时才需要实现
```
Comment (PaParaZz1) [inline source/04_best_practice/reward_model_zh.rst:43]: 加个举例，例如 Inverse RL 算法 Guided Cost Learning，T-REX

File: source/04_best_practice/reward_model_zh.rst [hunk 9]
```diff
@@ -0,0 +1,286 @@
+Reward Model 使用指南
+===============================
+
+Reward Model 入门
+-------------------------------
+
+**Reward Model 的基本概念**
```
Comment (PaParaZz1) [inline source/04_best_practice/reward_model_zh.rst:7]: 需要画一张图，表明 RM 在标准 RL pipeline 中的位置

File: source/04_best_practice/reward_model_zh.rst [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+    初始化包含的内容有，reward_model的config, reward model network, logger 和counter
+    self.cfg: 是config中的config.reward_model
+    self.reward_model: 我们自己写的Reward model Network
+    self.tb_logger: 目前使用的是tb_logger
+    self.counter: 各种你需要的计数器
+    """
+    pass
+
+    def collect_data(self, data: list) -> None:
+    """
```
Comment (PaParaZz1) [inline source/04_best_practice/reward_model_zh.rst:139]: 带上缩进吧

File: source/04_best_practice/reward_model_zh.rst [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+    这个方法的在entry的inner loop中调用
+    这个方法的用途是在coop-train的时候，向RM增加新的数据（不用于pretrain）
+    传入的data应该是一个由dict组成的list， 
+    每个dict需要包含（特殊情况请在注释中写明，推荐用assert确定在运算前）
+    {"obs": torch.tensor, "next_obs": torch.tensor, "action": torch.tensor, "reward": torch.tensor}
```
Comment (PaParaZz1) [inline source/04_best_practice/reward_model_zh.rst:144]: torch.Tensor

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]
